LPS-126597 Remember Device

martamedio opened the pull request:
Hey @csierra and @stian-sigvartsen 

Here's the code for the Remember Device issue that we have been discussing, so I'll try to summarize what we have done in this PR.

First of all, a new column has been added to know if the application has been flagged by the administrator to have a "_Remember Device_" option during authorization. 
If an application is marked as a "_Trusted Application_", then the "_Remember Device_" option won't show up at the admin screen, so an application cannot be "_Trusted Application_" and have the "_Remember Device_" option also checked.

We have also added a new column at the OAuth2Authorization level to associate the content of a cookie to each refresh token and a new filter to search by that cookie content.

The complexity is in the flow of OAuth2, we have overwritten several methods to intercept some requests, here is the process:

#### First time the user requests a token

1. User will request an authorization code (`/o/oauth2/authorize`) an option will be available to check if he/she wants to remember the device.
2. User wants to remember the device, when submitting the form (`/decision`) we track the checked option at `recreateRedirectionStateFromParams` and we add a cookie to the response + we add the value of the cookie as a property to track it during the next step.
3. User sends a request to obtain the token, at `LiferayAccessTokenService` we check if among the properties is the cookie content and we associate it to the new OAuth2Authorization entry.

#### Following times that the user requests a token (after having remembered the device)

1. User will request an authorization code (`/o/oauth2/authorize`) and the cookie will be present on the request, so at `canAuthorizationBeSkipped` we check the cookie value it's associated to a non-expired refresh token, we mark that token as expired and we allow to continue without manual authorization.
2. Before responding with the Authorization Code, at `getGrantRepresentation` we recover again the registry in which we have just invalidated the token, and we associate again the cookie value to the properties to use it in the next step.
3. User sends a request to obtain the token, at `LiferayAccessTokenService` we check if among the properties is the cookie content and we associate it to the new OAuth2Authorization entry.

Let me know what you think, it is not an easy functionality so be as critical as you can 🧐

liferay-appsec commented on modules/apps/oauth2-provider/oauth2-provider-api/src/main/resources/content/Language.properties:
```diff
@@ -4,6 +4,7 @@ CLIENT_CREDENTIALS=Client Credentials
 REFRESH_TOKEN=Refresh Token
```
Check that added properties are either titles or sentences

liferay-appsec commented on modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/admin/edit_application_left_column.jsp:
```diff
@@ -116,7 +116,7 @@ OAuth2Application oAuth2Application = oAuth2AdminPortletDisplayContext.getOAuth2
 							<c:choose>
 								<c:when test="<%= grantType.equals(GrantType.AUTHORIZATION_CODE) || grantType.equals(GrantType.AUTHORIZATION_CODE_PKCE) %>">
-									<aui:input checked="<%= checked %>" data="<%= data %>" label="<%= grantType.name() %>" name="<%= name %>" onchange='<%= liferayPortletResponse.getNamespace() + "updateTrustedApplicationSection();" %>' type="checkbox" />
+									<aui:input checked="<%= checked %>" data="<%= data %>" label="<%= grantType.name() %>" name="<%= name %>" onchange='<%= liferayPortletResponse.getNamespace() + "updateAdminOptionsApplicationSection();" %>' type="checkbox" />
```
This line adds a form input. Please check:

* The input is validated on the processing action

* The backend is protected agains value forging



Refer to documentation:

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

liferay-appsec commented on modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/admin/edit_application_left_column.jsp:
```diff
@@ -228,12 +228,25 @@ OAuth2Application oAuth2Application = oAuth2AdminPortletDisplayContext.getOAuth2
 				<h3 class="sheet-subtitle"><liferay-ui:message key="trusted-application" /></h3>
 
 				<aui:field-wrapper>
-					<aui:input checked="<%= (oAuth2Application == null) ? false : oAuth2Application.isTrustedApplication() %>" helpMessage="trusted-application-help" id="trustedApplication" label="trusted-application" name="trustedApplication" type="checkbox" />
```
This line adds a form input. Please check:

* The input is validated on the processing action

* The backend is protected agains value forging



Refer to documentation:

liferay-appsec commented on modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/admin/edit_application_left_column.jsp:
```diff
@@ -228,12 +228,25 @@ OAuth2Application oAuth2Application = oAuth2AdminPortletDisplayContext.getOAuth2
+			>
+				<h3 class="sheet-subtitle"><liferay-ui:message key="remember-device" /></h3>
+
+				<aui:field-wrapper>
```
This line adds a form input. Please check:

* The input is validated on the processing action

* The backend is protected agains value forging



Refer to documentation:

liferay-appsec commented on modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/authorize/authorize.jsp:
```diff
@@ -105,6 +105,12 @@ if (Validator.isNotNull(replyTo) && !replyTo.startsWith(PortalUtil.getPortalURL(
 
+						<c:if test="<%= oAuth2Application.isRememberDevice() %>">
+							<aui:field-wrapper>
+								<aui:input checked="<%= false %>" helpMessage="remember-device-help" id="rememberDevice" label="remember-device" name="rememberDevice" type="checkbox" />
```
This line adds a form input. Please check:

* The input is validated on the processing action

* The backend is protected agains value forging



Refer to documentation:

liferay-appsec commented on modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/content/Language.properties:
```diff
@@ -85,6 +85,9 @@ redirect-uris=Callback URIs
 redirect-uris-help=Specify where to redirect the user after the authorization.
```
Check that added properties are either titles or sentences

liferay-appsec reviewed (commented):
This is an automatic review suggestion

martamedio commented:
ci:test:sf

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-14.liferay.com/job/test-portal-source-format/4637/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-14.liferay.com//userContent/jobs/test-portal-source-format/builds/4637/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 3 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/7deedb562a6145be11b327fc35f9693b6b4e1541">7deedb562a6145be11b327fc35f9693b6b4e1541</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/liferay-appsec/liferay-portal/tree/LPS-126597">LPS-126597</a><br/>Branch GIT ID: <a href="https://github.com/liferay-appsec/liferay-portal/commit/72abfae3598e6f08a6ab751a5ea016d7ab04e07d">72abfae3598e6f08a6ab751a5ea016d7ab04e07d</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-14.liferay.com/job/test-portal-source-format/4637/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-14.liferay.com/userContent/jobs/test-portal-source-format/builds/4637/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<strong>Jenkins Build:</strong><a href="https://test-1-14.liferay.com/job/test-portal-source-format/4637/">test-portal-source-format#4637</a><br /><strong>Jenkins Report:</strong><a href="https://test-1-14.liferay.com/userContent/jobs/test-portal-source-format/builds/4637/jenkins-report.html">jenkins-report.html</a><br /><strong>Jenkins Suite:</strong> sf<br /><strong>Pull Request:</strong><a href="https://github.com/liferay-appsec/liferay-portal/pull/288">liferay-appsec#288</a><br /><strong>Spira Release:</strong><a href="https://liferay.spiraservice.net/16/Release/2002.aspx">/Liferay DXP 7.4/Pull Request/ci:test:sf</a><br /><strong>Spira Release Build:</strong><a href="https://liferay.spiraservice.net/16/Build/98732.aspx">master - liferay-appsec &gt; liferay-appsec - PR#288 - 2021-03-05[02:12:26]</a><br /><strong>Spira Jenkins Build:</strong><a href="https://test-1-16.liferay.com/job/publish-spira-report/17259/">publish-spira-report#17259</a><br />

achaparro commented on modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/internal/upgrade/OAuth2ServiceUpgrade.java:
```diff
@@ -73,7 +73,12 @@ public void register(Registry registry) {
 		registry.register(
 			"3.0.0", "4.0.0",
```
Hi @martamedio,

Shouldn't be from 3.0.0 to 3.1.0? Also, please remember to change the field Liferay-Require-SchemaVersion in bnd.bnd

Thanks.

martamedio replied:
Hey @achaparro, thanks for the heads up! 😃 

I'm keeping it in mind for when rebase from `master` (where it will already be: https://github.com/liferay-appsec/liferay-portal/pull/289)
This PR is still in process, so don't take this change into account at the moment, please, because all commits are previous to our conversation.

achaparro replied:
Sorry @martamedio, I didn't realize about that pull.

Thanks for the update

martamedio commented:
Hey @csierra and @stian-sigvartsen 

I've force push: rebase master + small fixes, so pick up the branch again, please.

I think the functionality is working, we should look at the cookie issue with a critical and security eyes 👀; and I'll need a hand to run a full test integration flow, I'm not sure how to do it without modifying the whole Base class.

Thanks guys!

stian-sigvartsen commented:
@martamedio Can we split up abfffc8 so it is more obvious what the actual change was? It looks like a mix of a change and source formatting?

martamedio commented:
Guys, I had to create a new branch because when I rebased the current one to fix conflicting files, I've messed up the whole commit history.

Please, see: https://github.com/liferay-appsec/liferay-portal/pull/296
In this one I've fixed the conflicts and I've also separated the commits as you mention 👇 
> @martamedio Can we split up [abfffc8](https://github.com/liferay-appsec/liferay-portal/commit/abfffc8c25ed16e6be070ba016803f4dfa87518a) so it is more obvious what the actual change was? It looks like a mix of a change and source formatting?

Thanks and sorry for any inconvenience!
